File: code/zato-web-admin/src/zato/admin/static/js/notif/cloud/openstack/swift.js

// /////////////////////////////////////////////////////////////////////////////

$.fn.zato.data_table.OpenstackSwiftNotification = new Class({
    toString: function() {
        var s = '<OpenstackSwiftNotification id:{0} name:{1} is_active:{2} is_snet:{3}>';
        return String.format(s, this.id ? this.id : '(none)',
                                this.name ? this.name : '(none)',
                                this.is_active ? this.is_active : '(none)'
                                );
    }
});

// /////////////////////////////////////////////////////////////////////////////

$(document).ready(function() {
    $('#data-table').tablesorter();
    $.fn.zato.data_table.class_ = $.fn.zato.data_table.OpenstackSwiftNotification;
    $.fn.zato.data_table.new_row_func = $.fn.zato.notif.cloud.openstack.swift.data_table.new_row;
    $.fn.zato.data_table.parse();
    $.fn.zato.data_table.setup_forms(['name', 'interval', 'containers', 'def_id', 'service_name']);
})

$.fn.zato.notif.cloud.openstack.swift.create = function() {
    $.fn.zato.data_table._create_edit('create', 'Create a new OpenStack Swift notification definition', null);
}

$.fn.zato.notif.cloud.openstack.swift.edit = function(id) {
    $.fn.zato.data_table._create_edit('edit', 'Update the OpenStack Swift notification definition', id);
}

$.fn.zato.notif.cloud.openstack.swift.data_table.new_row = function(item, data, include_tr) {
    var row = '';

    if(include_tr) {
        row += String.format("<tr id='tr_{0}' class='updated'>", item.id);
    }

    var is_active = item.is_active == true;
    var suppr_cons_slashes = item.suppr_cons_slashes == true;
    var name_pattern_neg = item.name_pattern_neg == true;

    var get_data_patt = item.get_data_patt ? item.get_data_patt : "";
    var get_data = item.get_data ? item.get_data : false;

    var name_pattern_neg = item.name_pattern_neg ? item.name_pattern_neg : false;
    var get_data_patt_neg = item.get_data_patt_neg ? item.get_data_patt_neg : false;

    var name_pattern = "<span class='form_hint'>(None)</span>";
    if(item.name_pattern) {
        name_pattern = item.name_pattern;
    }

    var containers_html = "";
    var containers = "";
    if(item.containers) {
        containers_html = item.containers;
        containers = item.containers;
    }

    console.log(item);

    row += "<td class='numbering'>&nbsp;</td>";
    row += "<td class='impexp'><input type='checkbox' /></td>";
    row += String.format('<td>{0}</td>', item.name);
    row += String.format('<td>{0}</td>', is_active ? 'Yes' : 'No');
    row += String.format('<td>{0}</td>', data.def_name);
    row += String.format('<td>{0}</td>', item.interval);
    row += String.format('<td>{0}</td>', name_pattern);
    row += String.format('<td>{0}</td>', name_pattern_neg ? 'Yes' : 'No');
    row += String.format('<td>{0}</td>', containers_html);

    row += String.format('<td>{0}</td>', String.format("<a href=\"javascript:$.fn.zato.notif.cloud.openstack.swift.edit('{0}')\">Edit</a>", item.id));
    row += String.format('<td>{0}</td>', String.format("<a href='javascript:$.fn.zato.notif.cloud.openstack.swift.delete_({0});'>Delete</a>", item.id));

    row += String.format("<td class='ignore item_id_{0}'>{0}</td>", item.id);
    row += String.format("<td class='ignore'>{0}</td>", is_active);
    row += String.format("<td class='ignore'>{0}</td>", item.def_id);

    row += String.format("<td class='ignore'>{0}</td>", name_pattern_neg);
    row += String.format("<td class='ignore'>{0}</td>", item.containers);

    row += String.format("<td class='ignore'>{0}</td>", get_data);
    row += String.format("<td class='ignore'>{0}</td>", get_data_patt);
    row += String.format("<td class='ignore'>{0}</td>", get_data_patt_neg);

    if(include_tr) {
        row += '</tr>';
    }

    return row;
}

$.fn.zato.notif.cloud.openstack.swift.delete_ = function(id) {
    $.fn.zato.data_table.delete_(id, 'td.item_id_',
        'OpenStack Swift notification definition [{0}] deleted',
        'Are you sure you want to delete the OpenStack Swift notification definition [{0}]?',
        true);
}

File: code/zato-web-admin/src/zato/admin/web/views/notif/cloud/openstack/swift.py
# -*- coding: utf-8 -*-

"""
Copyright (C) 2019, Zato Source s.r.o. https://zato.io

Licensed under LGPLv3, see LICENSE.txt for terms and conditions.
"""

from __future__ import absolute_import, division, print_function, unicode_literals

# stdlib
import logging

# Zato
from zato.admin.web.forms.notif.cloud.openstack.swift import CreateForm, EditForm
from zato.admin.web.views import CreateEdit, Delete as _Delete, Index as _Index
from zato.common.odb.model import NotificationOpenStackSwift as NotifOSS

logger = logging.getLogger(__name__)

common_required = ('name', 'is_active', 'def_id', 'containers', 'interval', 'name_pattern', 'name_pattern_neg', 'get_data',
    'get_data_patt_neg', 'service_name')

common_optional = ('get_data_patt',)

class Index(_Index):
    method_allowed = 'GET'
    url_name = 'notif-cloud-openstack-swift'
    template = 'zato/notif/cloud/openstack/swift.html'
    service_name = 'zato.notif.cloud.openstack.swift.get-list'
    output_class = NotifOSS
    paginate = True

    class SimpleIO(_Index.SimpleIO):
        input_required = ('cluster_id',)
        output_required = ('id', 'def_name',) + common_required
        output_optional = common_optional
        output_repeated = True

    def handle(self):

        def_list = []
        if self.req.zato.cluster_id:
            service_name = 'zato.cloud.openstack.swift.get-list'
            response = self.req.zato.client.invoke(service_name, {'cluster_id':self.req.zato.cluster_id})
            if response.has_data:
                def_list = response.data

        return {
            'create_form': CreateForm(def_list, req=self.req),
            'edit_form': EditForm(def_list, prefix='edit', req=self.req),
        }

class _CreateEdit(CreateEdit):
    method_allowed = 'POST'

    class SimpleIO(CreateEdit.SimpleIO):
        input_required = ('cluster_id',) + common_required
        input_optional = common_optional
        output_required = ('id', 'name', 'def_name')

    def success_message(self, item):
        return 'Successfully {0} the OpenStack Swift notification [{1}]'.format(self.verb, item.name)

class Create(_CreateEdit):
    url_name = 'notif-cloud-openstack-swift-create'
    service_name = 'zato.notif.cloud.openstack.swift.create'

class Edit(_CreateEdit):
    url_name = 'notif-cloud-openstack-swift-edit'
    form_prefix = 'edit-'
    service_name = 'zato.notif.cloud.openstack.swift.edit'

class Delete(_Delete):
    url_name = 'notif-cloud-openstack-swift-delete'
    error_message = 'Could not delete the OpenStack Swift notification'
    service_name = 'zato.notif.cloud.openstack.swift.delete'
